### ISSUE TITLE
setup.php should not assume a `.htaccess` file will exist and fail if it does not

### ISSUE THREAD

**Issue by AdamWill** [owncloud/core#14417]
In recent OC, `lib/private/setup.php` seems to have grown some code that assumes the OC install will have a `.htaccess` file and dies if it does not exist (and is not of a sufficiently recent 'version'), all to add some cutesy custom error documents:

	/**
	 * @return string Absolute path to htaccess
	 */
	private function pathToHtaccess() {
		return OC::$SERVERROOT.'/.htaccess';
	}

	/**
	 * Checks if the .htaccess contains the current version parameter
	 *
	 * @return bool
	 */
	private function isCurrentHtaccess() {
		$version = \OC_Util::getVersion();
		unset($version[3]);

		return !strpos(
			file_get_contents($this->pathToHtaccess()),
			'Version: '.implode('.', $version)
		) === false;
	}

	/**
	 * Append the correct ErrorDocument path for Apache hosts
	 *
	 * @throws \OC\HintException If .htaccess does not include the current version
	 */
	public static function updateHtaccess() {
		$setupHelper = new OC_Setup(\OC::$server->getConfig());
		if(!$setupHelper->isCurrentHtaccess()) {
			throw new \OC\HintException('.htaccess file has the wrong version. Please upload the correct version. Maybe you forgot to replace it after updating?');
		}

		$content = "\n";
		$content.= "ErrorDocument 403 ".OC::$WEBROOT."/core/templates/403.php\n";//custom 403 error page
		$content.= "ErrorDocument 404 ".OC::$WEBROOT."/core/templates/404.php";//custom 404 error page
		@file_put_contents($setupHelper->pathToHtaccess(), $content, FILE_APPEND); //suppress errors in case we don't have permissions for it
	}

This is not really a good idea. `.htaccess` files are usually considered a regrettable compromise for people on shared hosts who have limited administrative access to the host system, or for cases where you really need a site to be able to modify its own server configuration (which really isn't the best idea). Fedora puts all the appropriate configuration directives for ownCloud in a `/etc/http/conf.d/owncloud.conf` file - *including* the appropriate error document redirects! - and does not ship a `/usr/share/owncloud/.htaccess` (nor would OC be permitted to write to that file even if it did exist). I would expect other distro-packaged OC do something similar, and server admins deploying OC may well want to do the same.

There's already some much older stuff in `core/setup/controller.php` which messes with `.htaccess` files, but it's optional and contingent on a configuration directive, `check_for_working_htaccess` - I'm guessing that config directive was added for much the same reason.

Perhaps this code should also be skipped if `check_for_working_htaccess` is not true?

**Comment by LukasReschke** [comment]
See https://github.com/owncloud/core/issues/13693#issuecomment-75429739, this check is essential as without we had a ton of additional bug reports by users not properly updating their instances.

Should stay in my PoV.

**Comment by LukasReschke** [comment]
That said: Just disable the `.htaccess` in the Apache config and ship it anyways, I don't see how this is a problem?

**Comment by LukasReschke** [comment]
> check_for_working_htaccess - I'm guessing that config directive was added for much the same reason.

Actually that's used only to enable or disable the "is data directory public accessible"-security check.

**Comment by AdamWill** [comment]
"Actually that's used only to enable or disable the "is data directory public accessible"-security check."

Well sure, but it's all kinda the same thing, OC trying to get server configuration in line using `.htaccess` files. I kinda figured the reason it's configurable is exactly this situation.

"That said: Just disable the .htaccess in the Apache config and ship it anyways, I don't see how this is a problem?"

The problem is it then confuses people; if they see that file it's reasonable to assume it's there for a reason, then maybe they try and change something in it and it doesn't work...sure, I can stick big 'THIS IS JUST HERE TO MAKE OWNCLOUD CHECKS HAPPY' messages in it, but eh, the mess. Oh, well.

**Comment by LukasReschke** [comment]
Well, pull request welcome. I won't do that as I don't see much reason ;-)

**Comment by karlitschek** [comment]
@AdamWill As @LukasReschke said. We did this to do check if the upgrade is done correctly by users. We got a ton of bug reports that shown the people did mistakes which we want to detect here.
I totally agree that in a perfect world this wouldn´t be necessary. The configuration could be in the httpd.conf as you said and everything is perfect. But reality shows that this check seems to be necessary.

**Comment by AdamWill** [comment]
So that's why I'm suggesting to **LEAVE THE CHECK INCLUDED AND ENABLED BY DEFAULT** but just allow site admins / packagers who are handling the configuration in a different way to disable it!

I really don't how how many different ways I can say that. :)

**Comment by karlitschek** [comment]
@AdamWill O.K. I understand. Like @LukasReschke said a pull request would be welcome. I´m still not completely sure why a switch to disable this check is useful or why the .htaccess could do harm.
But it is totally fine to add a switch to disable this. :-)

**Comment by AdamWill** [comment]
it can't really do any harm to ship one that's just a dummy with whatever text makes OC happy, it's just a little confusing to a user poking around and it feels messy. maybe I'll just do that instead of sending a PR, I'll think about it.

**Comment by LukasReschke** [comment]
With 9.0 and the code checker we expect all files to be there as expected.
